tokio/src/runtime/basic_scheduler.rs

```diff
@@ -70,6 +72,9 @@ struct Shared {
 
     /// Unpark the blocked thread
     unpark: Box<dyn Unpark>,
+
+    // indicates whether the blocked on thread was woken
+    woken: AtomicBool,
```

Comment:
What happens if we have multiple calls to `block_on`?

Reply:
@Darksonn the field is set to false in `Spawner::waker_ref`, Wouldn't that work?

Reply:
If you implement the loom test I have mentioned below, that will answer this question.

Reply:
There is additional synchronization that ensures only one concurrent blocker hits this bool.